src/overlays/actors/ovl_Elf_Msg4/z_elf_msg4.c

```diff
@@ -30,30 +34,137 @@ const ActorInit Elf_Msg4_InitVars = {
 };
 
-#endif
-
-extern InitChainEntry D_80AFD8F0[];
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg4/D_80AFD900.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg4/func_80AFD380.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg4/ElfMsg4_Init.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg4/ElfMsg4_Destroy.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg4/func_80AFD5B4.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg4/func_80AFD5E0.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg4/func_80AFD668.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg4/func_80AFD770.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Elf_Msg4/ElfMsg4_Update.s")
+s32 func_80AFD380(ElfMsg4* this, PlayState* play) {
+    if ((this->actor.home.rot.y > 0) && (this->actor.home.rot.y < 0x81) &&
+        (Flags_GetSwitch(play, this->actor.home.rot.y - 1))) {
+        (void)"共倒れ"; // "Collapse together"
+        if (ELFMSG4_GET_SWITCHFLAG(&this->actor) != 0x7F) {
+            Flags_SetSwitch(play, ELFMSG4_GET_SWITCHFLAG(&this->actor));
+        }
+        Actor_MarkForDeath(&this->actor);
+        return true;
+    }
+    if (this->actor.home.rot.y == 0x81) {
+        if (Flags_GetClear(play, this->actor.room)) {
+            if (ELFMSG4_GET_SWITCHFLAG(&this->actor) != 0x7F) {
+                Flags_SetSwitch(play, ELFMSG4_GET_SWITCHFLAG(&this->actor));
+            }
+            Actor_MarkForDeath(&this->actor);
+            return true;
+        }
+    }
+    if (ELFMSG4_GET_SWITCHFLAG(&this->actor) == 0x7F) {
+        return false;
+    }
+    if (Flags_GetSwitch(play, ELFMSG4_GET_SWITCHFLAG(&this->actor))) {
+        (void)"共倒れ"; // "Collapse together"
+        Actor_MarkForDeath(&this->actor);
+        return true;
+    }
+    return false;
+}
+
+void ElfMsg4_Init(Actor* thisx, PlayState* play) {
+    ElfMsg4* this = THIS;
+
+    if (!func_80AFD380(this, play)) {
+        Actor_ProcessInitChain(&this->actor, sInitChain);
+
+        if (ABS_ALT(this->actor.home.rot.x) == 0) {
+            this->actor.scale.z = 0.4f;
+            this->actor.scale.x = 0.4f;
+        } else {
+            this->actor.scale.x = this->actor.scale.z = ABS_ALT(this->actor.home.rot.x) * 0.04f;
+        }
+        if (this->actor.home.rot.z == 0) {
+            this->actor.scale.y = 0.4f;
+        } else {
+            this->actor.scale.y = this->actor.home.rot.z * 0.04f;
+        }
+        this->actor.shape.rot.z = 0;
+        this->actionFunc = func_80AFD770;
+        this->elfMsg5 = NULL;
+        this->actor.shape.rot.x = this->actor.shape.rot.y = this->actor.shape.rot.z;
+    }
+}
+
+void ElfMsg4_Destroy(Actor* thisx, PlayState* play) {
+}
+
+s32 func_80AFD5B4(ElfMsg4* this) {
+    if (ELFMSG4_GET_8000(&this->actor) != 0) {
+        return ELFMSG4_GET_FF(&this->actor) + 0x200;
+    } else {
+        return -0x200 - ELFMSG4_GET_FF(&this->actor);
+    }
+}
```

Comment:
```suggestion
s32 func_80AFD5B4(ElfMsg4* this) {
    if (ELFMSG4_GET_8000(&this->actor)) {
        return ELFMSG4_GET_FF(&this->actor) + 0x200;
    } else {
        return  - ELFMSG4_GET_FF(&this->actor) - 0x200;
    }
}
```

This also matches and looks a bit clearer. Could call FF TEXT and the function `..._GetTextId` too.